File: CSharp - Monogame SDK 3.7/Asteroid.cs
using System;
using System.Collections.Generic;

namespace Asteroids {

    public class Asteroid : Entity {

        private Asteroid() { }

        public static void Spawn(Animation anim, double x, double y) {
            Queue<Asteroid> pool = new Queue<Asteroid>();

            Asteroid asteroid = (pool.Count == 0 || pool.Peek().active) ? new Asteroid() : pool.Dequeue();
            asteroid.anim = anim;
            asteroid.x = x;
            asteroid.y = y;
            asteroid.heading = Program.random.NextDouble() * 2 * Math.PI;
            asteroid.dx = Program.random.Next(1, 3); // move at a random speed
            asteroid.dy = Program.random.Next(1, 3);
            if (Program.random.Next(0, 1) == 0) asteroid.dx *= -1; // move to the left or right
            if (Program.random.Next(0, 1) == 0) asteroid.dy *= -1; // move up or down
            asteroid.active = true;

            pool.Enqueue(asteroid);
            Astro.Instance.entities.Add(asteroid);
        }

        override public void Update() {
            base.Update();
            base.Wrap();
        }

        override public bool CollidesWith(Entity e) {
            if (Intersects(e)) {
                active = false; // the asteroid gets destroyed
                Explosion.Spawn(Astro.Instance.animations["explosion_asteroid"], x, y);
                if (anim.r > 30) {
                    Astro.Instance.score += 10;
                    Astro.Instance.sounds["bang_lg"].Play();
                    Spawn(Astro.Instance.animations["asteroid_md"], x, y);
                    Spawn(Astro.Instance.animations["asteroid_md"], x, y);
                }
                else if (anim.r > 20) {
                    Astro.Instance.score += 20;
                    Astro.Instance.sounds["bang_md"].Play();
                    Spawn(Astro.Instance.animations["asteroid_sm"], x, y);
                    Spawn(Astro.Instance.animations["asteroid_sm"], x, y);
                }
                else {
                    Astro.Instance.score += 30;
                    Astro.Instance.sounds["bang_sm"].Play();
                }
                return true;
            }
            return false;
        }
    }
}

File: Java - AWT - Swing/asteroids/Asteroid.java
package asteroids;

import java.util.*;

class Asteroid extends Entity {

    private Asteroid() {
    }

    static public void spawn(Animation anim, float x, float y) {
        Queue<Asteroid> pool = new LinkedList<>();

        Asteroid asset = (pool.size() == 0 || pool.peek().active) ? new Asteroid() : pool.remove();
        asset.anim = anim;
        asset.x = x;
        asset.y = y;
        asset.heading = Astro.random(360);
        asset.dx = 1.0f + Astro.random(2); // move at a random speed
        asset.dy = 1.0f + Astro.random(2);
        if ((int) Astro.random(2) == 0)
            asset.dx *= -1; // move to the left or right
        if ((int) Astro.random(2) == 0)
            asset.dy *= -1; // move up or down
        asset.active = true;

        pool.add(asset);
        Astro.entities.add(0, asset);
    }

    public void update() {
        super.update();
        super.wrap();
    }

    public boolean collides_with(Entity e) {
        if (intersects(e)) {
            active = false; // the asteroid gets destroyed
            Explosion.spawn(Astro.animations.get("explosion_asteroid"), x, y);
            if (anim.r > 30) {
                Astro.play(Astro.sounds.get("bang_lg"));
                Asteroid.spawn(Astro.animations.get("asteroid_md"), x, y);
                Asteroid.spawn(Astro.animations.get("asteroid_md"), x, y);
            } else if (anim.r > 20) {
                Astro.play(Astro.sounds.get("bang_md"));
                Asteroid.spawn(Astro.animations.get("asteroid_sm"), x, y);
                Asteroid.spawn(Astro.animations.get("asteroid_sm"), x, y);
            } else {
                Astro.play(Astro.sounds.get("bang_sm"));
            }
            return true;
        }
        return false;
    }
}
